eppo_core/src/event_ingestion/retry.rs

```diff
@@ -1,40 +1,34 @@
 use crate::event_ingestion::batched_message::BatchedMessage;
-use crate::event_ingestion::delivery::QueuedBatch;
+use crate::event_ingestion::delivery::DeliveryStatus;
 use crate::event_ingestion::queued_event::QueuedEvent;
 use exponential_backoff::Backoff;
-use std::time::Duration;
 use log::warn;
+use std::time::Duration;
 use tokio::sync::mpsc;
 
 /// Retry events that failed to be delivered through `retry_downlink`, forwards remaining events to
 /// `delivery_status`.
 pub(super) async fn retry(
     mut uplink: mpsc::Receiver<BatchedMessage<QueuedEvent>>,
     retry_downlink: mpsc::Sender<BatchedMessage<QueuedEvent>>,
-    delivery_status: mpsc::Sender<QueuedBatch>,
     max_retries: u32,
     min_retry_duration: Duration,
     max_retry_delay: Duration,
 ) -> Option<()> {
     loop {
-        let BatchedMessage { batch, flush: _flush } = uplink.recv().await?;
-        if !batch.is_empty() {
+        let msg = uplink.recv().await?;
+        if !msg.batch.is_empty() {
             // take the number of attempts from the first event in the batch to determine the
             // exponential backoff delay
-            let attempts = batch[0].attempts as usize;
-            if wait_exponential_backoff(attempts, max_retries, min_retry_duration, max_retry_delay).await {
-                retry_downlink
-                    .send(BatchedMessage::new(batch.clone(), None))
-                    .await
-                    .ok()?;
-            } else {
+            let attempts = msg.batch[0].attempts as usize;
+            if !wait_exponential_backoff(attempts, max_retries, min_retry_duration, max_retry_delay)
+                .await
+            {
                 continue;
             }
+
+            retry_downlink.send(msg).await.ok()?;
         }
-        delivery_status
-            .send(QueuedBatch::retry(batch))
-            .await
-            .ok()?;
```

Comment:
Realized that it doesn't make sense for `retry` to send delivery statuses because `delivery` already sent those retries.

There was also a small bug here: `retry` was exiting if nobody was listening for delivery statuses (it should have ignored errors here)

This also allows us to avoid copying the batch (on line 27)

Reply:
nice find!